### ISSUE TITLE
Remote nodes do not trigger reconcile for MHC when MHC selector is empty

### ISSUE THREAD

**Issue by JoelSpeed** [kubernetes-sigs/cluster-api#3006]
**What steps did you take and what happened:**

Create an MHC with an empty selector
Add a condition to a node that is created by a Machine on the cluster the MHC targets

**What did you expect to happen:**
Expect MHC to be reconciled

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
When listing Machines that are targeted by MHC, an empty selector is allowed and matches all machines
https://github.com/kubernetes-sigs/cluster-api/blob/23a9ec341de33c399c542558c22046cb413082d5/controllers/machinehealthcheck_targets.go#L180-L197

When events are mapped from nodes to MHC, an empty selector causes the event to be dropped. This behaviour is therefore inconsistent with the above.
https://github.com/kubernetes-sigs/cluster-api/blob/23a9ec341de33c399c542558c22046cb413082d5/controllers/machine_helpers.go#L60-L63

**Environment:**

- Cluster-api version: master
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by ncdc** [comment]
I think we disallow empty selectors elsewhere (MachineDeployment/MachineSet)? Should we disallow empty selectors?

**Comment by JoelSpeed** [comment]
What's the motivation for disabling empty selectors elsewhere? I think an empty selector is valid for MachineHealthCheck, you may well want to health check over all machines with a single health check right? The checks we have in place already should make this safe.

I could see this either way tbh

**Comment by vincepri** [comment]
In the past we've gone through the same issue for MachineSet and MachineDeployment selectors, and we opted instead to default to some pre-defined label if the selector is empty. Folks brought up different issues with having a "catch-all" type of selector which might confuse users, and it'd be worthwhile to keep things consistent for in this project.

One example that comes to mind is, let's say soon we'll have KCP machines to be remediated, most users will probably want to setup a separate health check with different timeouts for control plane machines instead of using a global one. If you already have a MHC resource setup with an empty selector, and it matches everything you'll now automatically reconcile every machine including control plane ones.

To avoid users mistakes, and to always be explicit, we opted to validate that selectors should never be empty.

Hope this context helps clarify the historical question 🙂
